### ISSUE TITLE
Xamarin Forms Realm grows unexpectedly 

### ISSUE THREAD

**Issue by andrekiba** [realm/realm-dotnet#1634]
I have a Xamarin Forms app with realm 2.0.0, for now the only target is Android 7 api 24.
There is no realm platform (on server side there is sql server on windows with a web application which is the customer's software management).
I have implemented an asynchronous synchronization service based on rest api.
A sync request (push + pull) can be triggered from the ui (pull to refresh) or from android gcm scheduled task.
All calls are async and i have different threads (ConfigureAwait(False)).
I also have a singleton lazy RealmConfig like this:

```
private static readonly Lazy<RealmConfiguration> mobileConfig = new Lazy<RealmConfiguration>(() =>
        {
            const string dbPath = "hMobile.realm";
            var realmConfig = new RealmConfiguration(dbPath)
            {
                ShouldDeleteIfMigrationNeeded = false
            };            
            return realmConfig;
        });
public static RealmConfiguration MobileConfig => mobileConfig.Value;
```
Each time I need to read or write realm (so each time there is a thread switch) I call ``Realm.GetInstance()`` (and if I understood it well with version 2.0.0 the ``Realm.Refresh()`` is called automatically now) passing the same configuration.
And each time, even if I only call ``Realm.GetInstance()`` without do anything else (maybe there is no new data to sync) the size of the realm becomes double.
I expect that when a thread ends, its realm snapshot will be disposed correctly.
But it's not and I don't know why.
I need to explicitly call ``Realm.Dispose()`` ?

**Comment by nirinchev** [comment]
Yes, you need to explicitly call `Dispose` on your instances that are opened on background threads.

**Comment by andrekiba** [comment]
Ok, I try and let you know. Thanks!

**Comment by nirinchev** [comment]
Sounds good. I'm closing the ticket - if disposing doesn't work for you, feel free to reopen and we'll investigate further :)

**Comment by andrekiba** [comment]
Realm implements IDisposable so if I know it is a background thread without event loop can I do read/write actions inside a using statement and the Dispose method is automatically called?

**Comment by nirinchev** [comment]
Of course - that would be the optimal way to do it as your `Dispose` will be called even if an exception occurs.

**Comment by andrekiba** [comment]
Hi Nirinchev and thanks in advance. ``Dispose()`` seems to work as aspected but there is also another problem related to this. Realm grew very fast also because unnecessary synchronizations are triggered.
Just after a sync request (push + pull, so after the pull) the ``JobsLastSync `` is stored in app settings with code like this ``Settings.Current.JobsLastSync = DateTime.UtcNow;``
``Job`` is the entity i want to sync (it has a list of ``Node`` and some other properties, the one important in this case is ``LastUpdated``) and it overrides the ``OnPropertyChanged`` it this way:
```
protected override void OnPropertyChanged(string propertyName)
{
    if (!IsManaged || Realm.IsInTransaction || DoNotNotifyForTheseProperties.Contains(propertyName))
        return;
    if (propertyName == nameof(JobStatus))
    {
        RaisePropertyChanged(nameof(ToDo));
        RaisePropertyChanged(nameof(Active));
        RaisePropertyChanged(nameof(Completed));
    }

    Realm.Write(RefreshLastUpdate);
}

public void RefreshLastUpdate()
{
    LastUpdated = DateTimeOffset.UtcNow;
}
```
Inside the app user can do many things and as you can see each time some job's property changes the ``LastUpdated`` is automatically updated so the next sync can compare the ``LastUpdated`` of the job with the ``JobsLastSync`` and decide if something must be pushed or not.
The problem is that something calls ``OnPropertyChanged`` (4 times for each job's property) after the and of the pull and it is not a user action! So every time the job is is synched up again, even if there are no real changes.
The code of the sync method is something like this:
```
public async Task<bool> SyncJobs(bool fromSync = false, CancellationToken token = default(CancellationToken))
{
    //Debug.WriteLine($"Sync Jobs, before push => {Environment.CurrentManagedThreadId}");
    var realm = Realm.GetInstance(RealmConfig.HermesMobileConfig);   
    var lastSync = new DateTimeOffset(Settings.Current.JobsLastSync);
    var jobsToSync = realm.All<Job>().Where(j => j.LastUpdated >= lastSync 
                                                 || (j.Deleted != null && j.Deleted >= lastSync)
                                            );
    var pushOk = true;

    //dictionary  to store photos ids
    var nodePhotoIds = new Dictionary<int,List<string>>();

    if (jobsToSync.Any())
    {
        var jobsData = Mapper.Map<List<Job>, List<JobDTO>>(jobsToSync.ToList());
        jobsData.ForEach(j =>
        {
            j.Nodes.ForEach(n =>
            {
                if (!n.Photos.Any())
                    return;
                n.Photos.ForEach(p =>
                {
                    var bytes = storageService.ReadAllBytes(p.Path);
                    p.File = bytes;
                    p.Md5 = CryptoUtility.GetMd5Hash(bytes);
                });
                nodePhotoIds.Add(n.IdNodeJobItem, n.Photos.Select(x => x.IdPhoto).ToList());
            });
            
        });

        var pushJobsDto = new PushJobsDTO
        {
            IdUser = Settings.Current.IdUser,
            Jobs = jobsData
        };

        pushOk = await hermesService.PushJobsSync(pushJobsDto, fromSync, token).ConfigureAwait(false);
    }

    if (!pushOk)
        return false;

    var jobs = await hermesService.PullJobsSync(Settings.Current.IdVehicle, lastSync, fromSync, token).ConfigureAwait(false);

    if (jobs.Any())
    {
        //threads are different
        using (realm = Realm.GetInstance(RealmConfig.HermesMobileConfig))
        {
            realm.Write(() =>
            {
                jobs.ForEach(j =>
                {
                    var job = Mapper.Map<Job>(j);

                    //check if the job is new or not
                    //if the state is no longer correct for the app
                    //I have to delete it 
                    var exJob = realm.Find<Job>(job.IdJob);
                    var correctStatusList = Enum.GetValues(typeof(JobStatus)).Cast<byte>().ToList();
                    if (exJob != null && !correctStatusList.Contains(job.JobStatus))
                    {
                        exJob.Nodes.ForEach(realm.Remove);
                        realm.Remove(exJob);
                    }
                    else
                    {
                        var nodes = j.Nodes.Select(n =>
                        {
                            var additionals = Mapper.Map<List<Additional>>(n.Additionals);
                            var nodeTypes = Mapper.Map<List<NodeType>>(n.NodeTypes);
                            var node = Mapper.Map<Node>(n);
                            additionals.ForEach(a =>
                            {
                                realm.Add(a, true);
                                node.Additionals.Add(a);
                            });
                            nodeTypes.ForEach(nt =>
                            {
                                realm.Add(nt, true);
                                node.NodeTypes.Add(nt);
                            });

                            //restore photos
                            if (nodePhotoIds.ContainsKey(node.IdNodeJobItem))
                            {
                                var photoIds = nodePhotoIds[node.IdNodeJobItem];
                                photoIds.ForEach(id =>
                                {
                                    var photo = realm.Find<Photo>(id);
                                    if(photo != null)
                                        node.Photos.Add(photo);
                                });
                            }

                            return node;
                        });

                        nodes.ForEach(n =>
                        {
                            realm.Add(n, true);
                            job.Nodes.Add(n);
                        });

                        realm.Add(job, true);
                        job.AttachNodesNotifications();
                    }
                });
            });
        }
    }

    Settings.Current.JobsLastSync = DateTime.UtcNow;
    MessagingCenter.Send(App.Instance, SimpleMessages.JobsSynced);

    return true;
}
```

**Comment by nirinchev** [comment]
Realm will call `OnPropertyChanged` whenever the property is set (regardless if it's set to the same value or not). I'm not sure which class the overriden `OnPropertyChanged` method belongs to but if you're setting its properties (either directly or as a call to `Add(foo, update: true)`), the `OnPropertyChanged` method will be called. Even so, I'm not sure how/why this would be affecting file size growth.

**Comment by andrekiba** [comment]
Even if the object it's not yet managed? As you can see automapper `var job = Mapper.Map<Job>(j);` creates new `Job` with its list of `Node` (`Job` and `Node` extend `RealmObject`). At the end the job is added to the realm with update true because maybe it already exists.
Is there a way to attach to `PropertyChanged` only after the first creation/update?
If I call `AttachNotifications()` inside the constructor of the realm object
```
public void AttachNotifications()
{
    PropertyChanged += (sender, e) =>
    {
        if (!IsManaged || Realm.IsInTransaction || DoNotNotifyForTheseProperties.Contains(e.PropertyName))
            return;
        if (e.PropertyName == nameof(JobStatus))
        {
            RaisePropertyChanged(nameof(ToDo));
            RaisePropertyChanged(nameof(Active));
            RaisePropertyChanged(nameof(Completed));
        }
        Realm.Write(RefreshLastUpdate);
    };
}
```
is equal to override `PropertyChanged`, right?
And it works but I have the problem of the incorrect `LastUpdated`...so I thought to attach to the event after the object was constructed and added to realm...but in this case it doesn't work and the lamba is never called...

**Comment by nirinchev** [comment]
Yes, even unmanaged objects will get property change notifications but you already have the `if (!IsManaged)` check which will eliminate all invocations while the object is not yet managed.

**Comment by andrekiba** [comment]
Very thanks for your reply, I thought the same thing but that flag it's not useful in this case because unfortunately the invocation is called when the object is already managed...
Again, is there a way to attach to notifications only at some point in time and not inside constructor?

**Comment by nirinchev** [comment]
You can subscribe to notifications at any point, there's no need to subscribe in the constructor.

**Comment by andrekiba** [comment]
mmm...but, as mentioned earlier, if the subscription is outside of the constructor the lambda is never called...

**Comment by nirinchev** [comment]
That is highly unlikely as if that were true, then the Xamarin.Forms data bindings wouldn't work at all. Do you think you can make a small repro project that exhibits this behavior?

**Comment by andrekiba** [comment]
Not now Nikola but for sure I can share my desktop with you in debug when you have time.
I don't know why but if I call the above `AttachNotifications()` inside the constructor the lambda is called...otherwise not (and it does not depend on whether the object has already been added to relam or not).

**Comment by nirinchev** [comment]
I went over your code and it doesn't seem surprising that this happens. Because you dispose of the Realm as soon as you're done with it, this closes the Realm before it has a chance to notice that there were changes, therefore the property change notifications don't fire. You can force realm refresh by calling `realm.Refresh()` before disposing of it - this will advance the Realm to the latest version and fire any change notifications.

In general, I'd like to point out that this is a very fragile approach to update a property. There are several reasons for that:
1. On background threads, property notifications are not fired unless you explicitly refresh the Realm. 
2. `OnPropertyChanged` does not fire unless you have subscribed to `PropertyChanged`.
3. Disposing the Realm before notifications have had a chance to fire (what you're currently seeing) will cause these notifications to not fire.
4. Every time you get an object from the Realm, you get a new instance, so you'll need to subscribe for changes for every possible instance you get.

I'm not familiar with the architecture of your app, but a more solid approach would be to explicitly update the `LastUpdated` field whenever you manually make changes.

**Comment by andrekiba** [comment]
Thank you NikoIa, I start to understand now...during the sync process on background threads it's okay not to raise any `PropertyChanged`, it is exactly what I want.
But the most important thing here is the point 4, if I understood correctly every time I call `Realm.All` (for example) the constructor of each object is called again (and this is the reason of different behavior between outside or inside)...right?
I try to go on with your approach and explicitly update the `LastUpdate` manually...it seemed to me a good idea to not to have to remember it every time... :-(

**Comment by nirinchev** [comment]
Yes, every time you get an object, you'll get a different instance, e.g.

```csharp
var query = realm.All<Foo>();
var a = query.ElementAt(0);
var b = query.ElementAt(0);

// a and b are different instances
var expectedFalse = object.ReferenceEquals(a, b);

// but .Equals is overriden to check the underlying object in the database
var expectedTrue = a.Equals(b);
```

This essentially means any data you store in your object that is not persisted will not be there unless you pass the object by reference.

**Comment by andrekiba** [comment]
Thanks!
